src/restic_fmt_plug.c

```diff
@@ -0,0 +1,334 @@
+{
+	char *ctcopy = strdup(ciphertext);
+	char *keeptr = ctcopy;
+	int i;
+	char *p;
+	restic_salt *cur_salt;
+
+	cur_salt = mem_calloc_tiny(sizeof(restic_salt), MEM_ALIGN_SIMD);
+
+	ctcopy += TAG_LENGTH;
+
+	strtokm(ctcopy, "*"); /* SCRYPT */
+
+	p = strtokm(NULL, "*");
+	cur_salt->N = atoi(p);
+
+	p = strtokm(NULL, "*");
+	cur_salt->r = atoi(p);
+
+	p = strtokm(NULL, "*");
+	cur_salt->p = atoi(p);
+
+	p = strtokm(NULL, "*");
+	for (i = 0; i < sizeof(cur_salt->salt); i++)
+		cur_salt->salt[i] = atoi16[ARCH_INDEX(p[i * 2])] * 16 + atoi16[ARCH_INDEX(p[i * 2 + 1])];
+
+	p = strtokm(NULL, "*");
+	for (i = 0; i < sizeof(cur_salt->data); i++)
+		cur_salt->data[i] = atoi16[ARCH_INDEX(p[i * 2])] * 16 + atoi16[ARCH_INDEX(p[i * 2 + 1])];
+
+	MEM_FREE(keeptr);
+	return cur_salt;
+}
+
+unsigned int restic_tunable_cost_N(void *salt)
+{
+	restic_salt *rs = salt;
+	return (unsigned int)rs->N;
+}
+
+unsigned int restic_tunable_cost_r(void *salt)
+{
+	restic_salt *rs = salt;
+	return (unsigned int)rs->r;
+}
+
+unsigned int restic_tunable_cost_p(void *salt)
+{
+	restic_salt *rs = salt;
+	return (unsigned int)rs->p;
+}
+
+void restic_set_salt(void *salt)
+{
+	cur_salt = (restic_salt *)salt;
+}
+
+void restic_set_key(char *key, int index)
+{
+	strnzcpy(saved_key[index], key, sizeof(*saved_key));
+}
+char *restic_get_key(int index)
+{
+	return saved_key[index];
+}
+
+int restic_cmp_all(void *binary, int count)
+{
+	int index;
+
+	for (index = 0; index < count; index++)
+		if (((uint32_t *)binary)[0] == crypt_out[index][0])
+			return 1;
+	return 0;
+}
+
+int restic_cmp_one(void *binary, int index)
+{
+	return !memcmp(binary, crypt_out[index], BINARY_SIZE);
+}
+
+int restic_cmp_exact(char *source, int index)
+{
+	return 1;
+}
+
+static int restic_crypt_all(int *pcount, struct db_salt *salt)
+{
+	int count = *pcount;
+	int index;
+	int failed = 0;
+	yescrypt_params_t params = {.N = cur_salt->N, .r = cur_salt->r, .p = cur_salt->p};
+#ifdef _OPENMP
+#pragma omp parallel for
+#endif
+	for (index = 0; index < count; index += MIN_KEYS_PER_CRYPT) {
+		uint8_t kdf_out[64];
+#ifdef _OPENMP
+		int t = omp_get_thread_num();
+		if (t >= max_threads) {
+#pragma omp atomic
+			failed |= 2;
+			continue;
+		}
+#else
+		const int t = 0;
+#endif
+		if (yescrypt_kdf(NULL, &local[t], (const uint8_t *)saved_key[index], strlen(saved_key[index]),
+				 (const uint8_t *)cur_salt->salt, sizeof(cur_salt->salt), &params, kdf_out,
+				 sizeof(kdf_out))) {
+			failed |= 1;
+		}
+		uint8_t *poly1305_key = kdf_out + 32;
+		const unsigned char *nonce = cur_salt->data;
+		const unsigned char *ciphertext = cur_salt->data + NONCE_SIZE;
+		unsigned char prepared_key[32];
+		unsigned char key_mask[16] = {0xff, 0xff, 0xff, 0x0f, 0xfc, 0xff, 0xff, 0x0f,
+					      0xfc, 0xff, 0xff, 0x0f, 0xfc, 0xff, 0xff, 0x0f};
+		unsigned char masked_key[32];
+		int i;
+		memcpy(masked_key, poly1305_key, 32);
+		for (i = 0; i < 16; i++) {
+			masked_key[i + 16] &= key_mask[i];
+		}
```

Comment:
It should be possible to simplify and optimize this to two `uint64_t` ANDs (with corresponding changes to types and thus alignment of these variables), not needing a loop.

Reply:
:+1:  Indeed: Done!

Initial code was just a simple/naive port of the corresponding GOLANG code.